File: frontend/src/types/structure.ts
export interface StructureSource {
  id: string
  parser: string
  apiBase: string
}

export interface AseFrame {
  frame_index: number
  positions: number[][]
  cell: number[][]
  pbc: boolean[]
  tags: number[]
  fixed_indices: number[]
  energy?: number | null
  fmax?: number | null
  symbols?: string[] | null
  numbers?: number[] | null
}

export interface AsePreviewResponse {
  path: string
  name: string
  format?: string | null
  transport: 'json' | 'binary-available'
  is_trajectory: boolean
  n_frames: number
  n_atoms: number
  initial_frame_index: number
  topology_stable: boolean
  size_limit_overridden?: boolean
  frame: AseFrame
  source?: StructureSource
}

export interface BinaryArraySpec {
  offset: number
  byte_length: number
  shape: number[]
}

export interface AseFrameChunkHeader {
  version: number
  dtype: 'float32-le'
  int_dtype: 'int32-le'
  start: number
  count: number
  n_frames_total: number
  n_atoms: number
  topology_stable: boolean
  path?: string
  name?: string
  format?: string | null
  symbols: string[]
  numbers: number[]
  pbc: boolean[]
  arrays: Record<string, BinaryArraySpec>
  nan_means_null?: string[]
}

export interface AseFrameChunk {
  header: AseFrameChunkHeader
  buffer: ArrayBuffer
  dataStart: number
  positions: Float32Array
  cells: Float32Array
  tags?: Int32Array
  fixedMask?: Uint8Array
  energy?: Float32Array
  fmax?: Float32Array
}

export interface AseFrameJsonChunk {
  start: number
  count: number
  frames: AseFrame[]
}

File: backend/app/models/structure.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class AseFrame(BaseModel):
    frame_index: int
    positions: list[list[float]]
    cell: list[list[float]]
    pbc: list[bool]
    tags: list[int]
    fixed_indices: list[int] = Field(default_factory=list)
    energy: Optional[float] = None
    fmax: Optional[float] = None
    symbols: Optional[list[str]] = None
    numbers: Optional[list[int]] = None


class AsePreviewResponse(BaseModel):
    path: str
    name: str
    format: Optional[str] = None
    transport: Literal["json", "binary-available"]
    is_trajectory: bool
    n_frames: int
    n_atoms: int
    initial_frame_index: int
    topology_stable: bool = True
    size_limit_overridden: bool = False
    frame: AseFrame
    file_incomplete: bool = False
    scan_completed: bool = True
    warnings: list[str] = Field(default_factory=list)


class AseFrameChunkResponse(BaseModel):
    start: int
    count: int
    frames: list[AseFrame]
